### ISSUE TITLE
Rendering for custom multicolumn expectations failing

### ISSUE THREAD

**Issue by ragarciafran** [fivetran/great_expectations#6937]
**Describe the bug**
I defined a custom expectation which seems to be working based on unit tests and validation results, but when rendering in data docs it shows up kinda funky (see image). I defined the renderer following a couple of examples and as far as I can tell it’s correct. There’s no traceback anywhere I can see. Any help with this would be much appreciated
GX version 0.15.26
Debian GNU/Linux 11 (bullseye)

![image](https://user-images.githubusercontent.com/47935089/214926804-e6aaf33e-55e2-4799-92c9-5e1074e10f0f.png)

Code for custom expectation:

```
"""
This is a template for creating custom MulticolumnMapExpectations.
For detailed instructions on how to use it, please see:
    https://docs.greatexpectations.io/docs/guides/expectations/creating_custom_expectations/how_to_create_custom_multicolumn_map_expectations
"""

from typing import Optional

from great_expectations.core.expectation_configuration import ExpectationConfiguration
from great_expectations.exceptions import InvalidExpectationConfigurationError
from great_expectations.execution_engine import (
    PandasExecutionEngine,
    SqlAlchemyExecutionEngine,
)
from great_expectations.expectations.expectation import MulticolumnMapExpectation
from great_expectations.expectations.metrics.map_metric_provider import (
    MulticolumnMapMetricProvider,
    multicolumn_condition_partial,
)
from great_expectations.expectations.util import render_evaluation_parameter_string
from great_expectations.render.renderer.renderer import renderer
from great_expectations.render.types import RenderedStringTemplateContent
from great_expectations.render.util import (
    handle_strict_min_max,
    substitute_none_for_missing,
)
from sqlalchemy.sql import func


# This class defines a Metric to support your Expectation.
# For most MulticolumnMapExpectations, the main business logic for calculation will live in this class.
# <snippet>
class MulticolumnValuesSumBetweenAAndB(MulticolumnMapMetricProvider):
    # </snippet>
    # This is the id string that will be used to reference your metric.
    # <snippet>
    condition_metric_name = "multicolumn_values.sum_between_a_and_b"
    # </snippet>
    # These point your metric at the provided keys to facilitate calculation
    condition_domain_keys = (
        "batch_id",
        "table",
        "column_list",
        "row_condition",
        "condition_parser",
        "ignore_row_if",
    )
    condition_value_keys = ("min_value", "strict_min", "max_value", "strict_max")

    # This method implements the core logic for the PandasExecutionEngine
    # <snippet>
    @multicolumn_condition_partial(engine=PandasExecutionEngine)
    def _pandas(cls, column_list, **kwargs):
        raise NotImplementedError

    # </snippet>

    # This method defines the business logic for evaluating your metric when using a SqlAlchemyExecutionEngine
    @multicolumn_condition_partial(engine=SqlAlchemyExecutionEngine)
    def _sqlalchemy(cls, column_list, **kwargs):
        # Obtaining components needed for validation
        min_value = kwargs.get("min_value")
        strict_min = kwargs.get("strict_min")
        max_value = kwargs.get("max_value")
        strict_max = kwargs.get("strict_max")

        column_list_0_filled = [func.ifnull(col, 0) for col in column_list]

        column_sum = sum(column_list_0_filled)

        if min_value is not None:
            if strict_min:
                above_min = column_sum > min_value
            else:
                above_min = column_sum >= min_value
        else:
            above_min = None

        if max_value is not None:
            if strict_max:
                below_max = column_sum < max_value
            else:
                below_max = column_sum <= max_value
        else:
            below_max = None

        # not a huge fan of this, but I couldn't figure out
        # how to do it otherwise.
        if above_min is None:
            return below_max
        elif below_max is None:
            return above_min
        else:
            return above_min & below_max

    # This method defines the business logic for evaluating your metric when using a SparkDFExecutionEngine
    # @multicolumn_condition_partial(engine=SparkDFExecutionEngine)
    # def _spark(cls, column_list, **kwargs):
    #     raise NotImplementedError


# This class defines the Expectation itself
# <snippet>
class ExpectMulticolumnValuesSumBetweenAAndB(MulticolumnMapExpectation):
    # </snippet>
    # <snippet>
    """Expect the sum of a list of columns to be between a and b.
    Treats NULLs as 0.
    args:
        - column_list: list of columns that will be added up.
    kwargs:
        - min_value: lower bound. Defaults to None, which removes the lower bound.
        - max_value: upper bound. Defaults to None, which removes the upper bound.
        Note that min_value and max_value cannot both be None.
        - strictly_min: bool, determines whether it can also equal the lower bound.
        Defaults to False.
        - strictly_max: bool, determines whether it can also equal the upper bound.
        Defaults to False.
    """
    # </snippet>

    # These examples will be shown in the public gallery.
    # They will also be executed as unit tests for your Expectation.
    examples = [
        {
            "data": {
                "col_a": [0.1, 0.2, 0.3, 0.4, 0.5],
                "col_b": [0.5, 0.1, 0.3, 0.2, 0.5],
                "col_c": [None, 0.1, None, 0.2, None],
                "col_d": [1, 1, 1, 1, 1],
                "col_e": [-1, -1, -1, -1, -1],
                "col_f": [1, 0, 1, 0, 1],
            },
            "tests": [
                {
                    "title": "basic_positive_test",
                    "exact_match_out": False,
                    "include_in_gallery": True,
                    "in": {
                        "column_list": ["col_a", "col_b"],
                        "min_value": 0,
                        "max_value": 1,
                        "strict_min": False,
                        "strict_max": False,
                    },
                    "out": {
                        "success": True,
                    },
                },
                {
                    "title": "basic_negative_test",
                    "exact_match_out": False,
                    "include_in_gallery": True,
                    "in": {
                        "column_list": ["col_a", "col_b", "col_d"],
                        "min_value": 0,
                        "max_value": 1,
                        "strict_min": False,
                        "strict_max": False,
                    },
                    "out": {
                        "success": False,
                    },
                },
                {
                    "title": "positive_test_with_NULLs",
                    "exact_match_out": False,
                    "include_in_gallery": True,
                    "in": {
                        "column_list": ["col_a", "col_c"],
                        "min_value": 0,
                        "max_value": 1,
                        "strict_min": False,
                        "strict_max": False,
                    },
                    "out": {
                        "success": True,
                    },
                },
                {
                    "title": "negative_test_with_NULLs",
                    "exact_match_out": False,
                    "include_in_gallery": True,
                    "in": {
                        "column_list": ["col_a", "col_c", "col_f"],
                        "min_value": 0,
                        "max_value": 1,
                        "strict_min": False,
                        "strict_max": False,
                    },
                    "out": {
                        "success": False,
                    },
                },
                {
                    "title": "positive_test_without_lower_bound",
                    "exact_match_out": False,
                    "include_in_gallery": True,
                    "in": {
                        "column_list": ["col_a", "col_e"],
                        "min_value": None,
                        "max_value": 1,
                        "strict_min": False,
                        "strict_max": False,
                    },
                    "out": {
                        "success": True,
                    },
                },
            ],
            "test_backends": [
                {
                    "backend": "sqlalchemy",
                    "dialects": ["sqlite", "postgresql"],
                },
            ],
        }
    ]

    # This is the id string of the Metric used by this Expectation.
    # For most Expectations, it will be the same as the `condition_metric_name` defined in your Metric class above.
    # <snippet>
    map_metric = "multicolumn_values.sum_between_a_and_b"
    # </snippet>

    # This is a list of parameter names that can affect whether the Expectation evaluates to True or False
    success_keys = ("min_value", "strict_min", "max_value", "strict_max")

    # This dictionary contains default values for any parameters that should have default values
    default_kwarg_values = {
        "min_value": None,
        "max_value": None,
        "strict_min": False,
        "strict_max": False,
    }

    args_keys = (
        "column_list",
        "min_value",
        "max_value",
        "strict_min",
        "strict_max",
    )

    def validate_configuration(
        self, configuration: Optional[ExpectationConfiguration] = None
    ) -> None:
        """
        Validates that a configuration has been set, and sets a configuration if it has yet to be set. Ensures that
        necessary configuration arguments have been provided for the validation of the expectation.
        Args:
            configuration (OPTIONAL[ExpectationConfiguration]): \
                An optional Expectation Configuration entry that will be used to configure the expectation
        Returns:
            None. Raises InvalidExpectationConfigurationError if the config is not validated successfully
        """

        super().validate_configuration(configuration)
        configuration = configuration or self.configuration

        # # Check other things in configuration.kwargs and raise Exceptions if needed
        min_value = configuration.kwargs["min_value"]
        max_value = configuration.kwargs["max_value"]
        strict_min = configuration.kwargs["strict_min"]
        strict_max = configuration.kwargs["strict_max"]

        # Validating that min_val, max_val, strict_min, and strict_max are of the proper format and type
        try:
            assert (
                min_value is not None or max_value is not None
            ), "min_value and max_value cannot both be none"
            assert min_value is None or isinstance(
                min_value, (float, int)
            ), "Provided min threshold must be a number"
            assert max_value is None or isinstance(
                max_value, (float, int)
            ), "Provided max threshold must be a number"
            if min_value and max_value:
                assert (
                    min_value <= max_value
                ), "Provided min threshold must be less than or equal to max threshold"
            assert strict_min is None or isinstance(
                strict_min, bool
            ), "strict_min must be a boolean value"
            assert strict_max is None or isinstance(
                strict_max, bool
            ), "strict_max must be a boolean value"
        except AssertionError as e:
            raise InvalidExpectationConfigurationError(str(e))

    @classmethod
    def _atomic_prescriptive_template(
        cls,
        configuration=None,
        result=None,
        language=None,
        runtime_configuration=None,
        **kwargs,
    ):
        runtime_configuration = runtime_configuration or {}
        include_column_name = runtime_configuration.get("include_column_name", True)
        include_column_name = (
            include_column_name if include_column_name is not None else True
        )
        styling = runtime_configuration.get("styling")
        params = substitute_none_for_missing(
            configuration.kwargs,
            [
                "column_list",
                "min_value",
                "max_value",
                "strict_min",
                "strict_max",
            ],
        )
        params_with_json_schema = {
            "column_list": {
                "schema": {"type": "string"},
                "value": params.get("column_list"),
            },
            "min_value": {
                "schema": {"type": "number"},
                "value": params.get("min_value"),
            },
            "max_value": {
                "schema": {"type": "number"},
                "value": params.get("max_value"),
            },
            "strict_min": {
                "schema": {"type": "boolean"},
                "value": params.get("strict_min"),
            },
            "strict_max": {
                "schema": {"type": "boolean"},
                "value": params.get("strict_max"),
            },
        }

        at_least_str, at_most_str = handle_strict_min_max(params)

        if params["min_value"] is not None and params["max_value"] is not None:
            template_str = f"Sum must be {at_least_str} $min_value and {at_most_str} $max_value for these columns: "
        elif params["min_value"] is None:
            template_str = f"Sum must be {at_most_str} $max_value for these columns: "
        elif params["max_value"] is None:
            template_str = f"Sum must be {at_least_str} $min_value for these columns: "

        for idx in range(len(params["column_list"]) - 1):
            template_str += f"$column_list_{str(idx)}, "
            params[f"column_list_{str(idx)}"] = params["column_list"][idx]

        last_idx = len(params["column_list"]) - 1
        template_str += f"$column_list_{str(last_idx)}"
        params[f"column_list_{str(last_idx)}"] = params["column_list"][last_idx]

        return (template_str, params_with_json_schema, styling)

    @classmethod
    @renderer(renderer_type="renderer.prescriptive")
    @render_evaluation_parameter_string
    def _prescriptive_renderer(
        cls,
        configuration=None,
        result=None,
        language=None,
        runtime_configuration=None,
        **kwargs,
    ):
        runtime_configuration = runtime_configuration or {}
        include_column_name = runtime_configuration.get("include_column_name", True)
        include_column_name = (
            include_column_name if include_column_name is not None else True
        )
        styling = runtime_configuration.get("styling")
        params = substitute_none_for_missing(
            configuration.kwargs,
            [
                "column_list",
                "min_value",
                "max_value",
                "strict_min",
                "strict_max",
            ],
        )

        at_least_str, at_most_str = handle_strict_min_max(params)

        if params["min_value"] is not None and params["max_value"] is not None:
            template_str = f"Sum must be {at_least_str} $min_value and {at_most_str} $max_value for these columns: "
        elif params["min_value"] is None:
            template_str = f"Sum must be {at_most_str} $max_value for these columns: "
        elif params["max_value"] is None:
            template_str = f"Sum must be {at_least_str} $min_value for these columns: "

        for idx in range(len(params["column_list"]) - 1):
            template_str += f"$column_list_{str(idx)}, "
            params[f"column_list_{str(idx)}"] = params["column_list"][idx]

        last_idx = len(params["column_list"]) - 1
        template_str += f"$column_list_{str(last_idx)}"
        params[f"column_list_{str(last_idx)}"] = params["column_list"][last_idx]

        return [
            RenderedStringTemplateContent(
                **{
                    "content_block_type": "string_template",
                    "string_template": {
                        "template": template_str,
                        "params": params,
                        "styling": styling,
                    },
                }
            )
        ]

    @classmethod
    @renderer(renderer_type="renderer.diagnostic.observed_value")
    def _diagnostic_observed_value_renderer(
        cls,
        configuration=None,
        result=None,
        language=None,
        runtime_configuration=None,
        **kwargs,
    ) -> None:
        pass

    # This object contains metadata for display in the public Gallery
    # <snippet>
    library_metadata = {
        "tags": [],  # Tags for this Expectation in the Gallery
        "contributors": [  # Github handles for all contributors to this Expectation.
            "@ragarciafran",  # Don't forget to add your github handle here!
        ],
    }
    # </snippet>


if __name__ == "__main__":
    # <snippet>
    ExpectMulticolumnValuesSumBetweenAAndB().print_diagnostic_checklist()
# </snippet>
```

**To Reproduce**
1. Define custom expectation (see code above)
2. Go through the typical GX workflow and build data docs
3. Examine data docs

**Expected behavior**
Results rendering correctly (as in the top of the image for a separate set of expectations. 

**Environment (please complete the following information):**
GX version 0.15.26
Debian GNU/Linux 11 (bullseye)

**Comment by talagluck** [comment]
Hi @ragarciafran - thanks for raising! We will review and be in touch. In the meantime, would you try defining the diagnostic renderer as well (similar to https://github.com/great-expectations/great_expectations/blob/605eb869c8ec011ba6c3a9609b5e07419d378932/tests/integration/docusaurus/expectations/creating_custom_expectations/expect_column_values_to_equal_three.py#L150). Currently, your method just has a `pass`.

**Comment by ragarciafran** [comment]
Hi there, I defined the diagnostic renderer following the example you set. Not sure if you hoped that would produce a change in the rendering behavior, but in any case it did not. 

Let me know if anything comes up.

**Comment by talagluck** [comment]
Hi @ragarciafran - was this resolved for you by #7041 ?
